File: android/src/main/java/ee/forgr/capacitor/social/login/XProvider.java
package ee.forgr.capacitor.social.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Base64;
import android.util.Log;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import ee.forgr.capacitor.social.login.helpers.SocialProvider;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.concurrent.CompletableFuture;

public class XProvider implements SocialProvider {
    private static final String TAG = "XProvider";
    private static final String OAUTH_URL = "https://x.com/i/oauth2/authorize";
    private static final String TOKEN_URL = "https://api.x.com/2/oauth2/token";
    private static final String USER_PROFILE_URL = "https://api.x.com/2/users/me";
    private static final int CUSTOM_TAB_REQUEST_CODE = 1001;
    
    private Activity activity;
    private Context context;
    private String clientId;
    private String redirectUrl;
    private String accessToken;
    private String refreshToken;
    private PluginCall currentCall;
    private boolean isHandlingOAuth = false; // Flag to prevent race condition
    
    public XProvider(Activity activity, Context context) {
        this.activity = activity;
        this.context = context;
    }
    
    public void initialize(String clientId, String redirectUrl) {
        this.clientId = clientId;
        this.redirectUrl = redirectUrl;
        // loadStoredTokens();
    }
    
    @Override
    public void login(PluginCall call, JSONObject config) {
        if (clientId == null || clientId.isEmpty()) {
            call.reject("X Client ID not set. Call initialize() first.");
            return;
        }
        
        try {
            // Generate PKCE parameters
            String codeVerifier = generateCodeVerifier();
            String codeChallenge = generateCodeChallenge(codeVerifier);
            String state = generateState();
            
            // Store code verifier for later use
            storeCodeVerifier(codeVerifier);
            
            // Build OAuth URL
            String scopes = "tweet.read users.read offline.access";
            if (config.has("scopes")) {
                scopes = config.getString("scopes");
            }
            
            String authUrl = buildAuthUrl(codeChallenge, state, scopes);
            
            // Show in-app browser
            showInAppBrowser(authUrl, call);
            
        } catch (Exception e) {
            Log.e(TAG, "Error during login", e);
            call.reject("Login failed: " + e.getMessage());
        }
    }
    
    @Override
    public void logout(PluginCall call) {
        clearCodeVerifier();
        call.resolve();
    }
    
    @Override
    public void getAuthorizationCode(PluginCall call) {
        if (accessToken == null || accessToken.isEmpty()) {
            call.reject("No X authorization code available");
            return;
        }
        
        JSObject result = new JSObject();
        result.put("accessToken", accessToken);
        call.resolve(result);
    }
    
    @Override
    public void isLoggedIn(PluginCall call) {
        // For X provider, we don't store tokens locally
        // Backend should handle token validation
        JSObject result = new JSObject();
        result.put("isLoggedIn", false);
        call.resolve(result);
    }
    
    @Override
    public void refresh(PluginCall call) {
        call.reject("Refresh not implemented for X provider");
    }
    
    private String generateCodeVerifier() {
        SecureRandom secureRandom = new SecureRandom();
        byte[] codeVerifier = new byte[32];
        secureRandom.nextBytes(codeVerifier);
        return Base64.encodeToString(codeVerifier, Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP);
    }
    
    private String generateCodeChallenge(String codeVerifier) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hash = digest.digest(codeVerifier.getBytes(StandardCharsets.UTF_8));
        return Base64.encodeToString(hash, Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP);
    }
    
    private String generateState() {
        SecureRandom secureRandom = new SecureRandom();
        byte[] state = new byte[16];
        secureRandom.nextBytes(state);
        return Base64.encodeToString(state, Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP);
    }
    
    private String buildAuthUrl(String codeChallenge, String state, String scopes) {
        Uri.Builder builder = Uri.parse(OAUTH_URL).buildUpon();
        builder.appendQueryParameter("response_type", "code");
        builder.appendQueryParameter("client_id", clientId);
        builder.appendQueryParameter("redirect_uri", redirectUrl);
        builder.appendQueryParameter("scope", scopes);
        builder.appendQueryParameter("state", state);
        builder.appendQueryParameter("code_challenge", codeChallenge);
        builder.appendQueryParameter("code_challenge_method", "S256");
        
        return builder.build().toString();
    }
    
    private void showInAppBrowser(String authUrl, PluginCall call) {
        this.currentCall = call;
        
        activity.runOnUiThread(() -> {
            try {
                // Create Custom Tabs Intent
                CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
                builder.setToolbarColor(ContextCompat.getColor(context, android.R.color.white));
                builder.setShowTitle(true);
                builder.setUrlBarHidingEnabled(false);
                
                CustomTabsIntent customTabsIntent = builder.build();
                customTabsIntent.intent.setData(Uri.parse(authUrl));
                
                // Launch Custom Tab
                activity.startActivity(customTabsIntent.intent);
                
                // Set timeout to handle case when user closes tab manually
                new android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(() -> {
                    if (currentCall != null) {
                        Log.w(TAG, "OAuth timeout - user may have closed the tab manually");
                        currentCall.reject("OAuth cancelled by user");
                        currentCall = null;
                        clearCodeVerifier();
                    }
                }, 300000); // 5 minutes timeout
                
            } catch (Exception e) {
                Log.e(TAG, "Error launching Custom Tab", e);
                call.reject("Failed to launch browser: " + e.getMessage());
                currentCall = null;
            }
        });
    }
    
    private void handleOAuthCallback(String url) {
        if (currentCall == null || isHandlingOAuth) {
            Log.e(TAG, "No current call available or already handling OAuth");
            return;
        }
        
        isHandlingOAuth = true;
        
        Uri uri = Uri.parse(url);
        String code = uri.getQueryParameter("code");
        String state = uri.getQueryParameter("state");
        String error = uri.getQueryParameter("error");
        
        if (error != null) {
            Log.e(TAG, "OAuth error: " + error);
            currentCall.reject("OAuth error: " + error);
            currentCall = null;
            isHandlingOAuth = false;
            return;
        }
        
        if (code != null && state != null) {
            // Return code and code_verifier for backend to exchange
            String codeVerifier = getStoredCodeVerifier();
            if (codeVerifier == null) {
                currentCall.reject("No code verifier found");
                currentCall = null;
                isHandlingOAuth = false;
                return;
            }
            
            // Clear code verifier
            clearCodeVerifier();
            
            // Return result
            JSObject result = new JSObject();
            result.put("provider", "x");
            
            JSObject resultData = new JSObject();
            resultData.put("token", code);
            resultData.put("code_verifier", codeVerifier);
            
            result.put("result", resultData);
            
            currentCall.resolve(result);
            currentCall = null;
            isHandlingOAuth = false;
        } else {
            currentCall.reject("Invalid OAuth callback");
            currentCall = null;
            isHandlingOAuth = false;
        }
    }
    
    public boolean handleIntent(Intent intent) {
        if (intent != null && intent.getData() != null) {
            String url = intent.getData().toString();
            if (url.startsWith(redirectUrl)) {
                handleOAuthCallback(url);
                return true;
            }
        }
        return false;
    }
    
    /**
     * Handle app resume - check if OAuth was cancelled by user
     */
    public void onAppResume() {
        // If we have a current call but no redirect happened, user probably cancelled
        if (currentCall != null && !isHandlingOAuth) {
            // Check if we have a stored code verifier (meaning OAuth was started)
            String codeVerifier = getStoredCodeVerifier();
            if (codeVerifier != null) {
                // User probably closed the tab manually, clear everything
                Log.w(TAG, "App resumed with pending OAuth call - user may have cancelled");
                currentCall.reject("OAuth cancelled by user");
                currentCall = null;
                clearCodeVerifier();
            }
        }
    }

    private void storeCodeVerifier(String codeVerifier) {
        android.content.SharedPreferences prefs = context.getSharedPreferences("XProvider", Context.MODE_PRIVATE);
        prefs.edit().putString("code_verifier", codeVerifier).apply();
    }
    
    private String getStoredCodeVerifier() {
        android.content.SharedPreferences prefs = context.getSharedPreferences("XProvider", Context.MODE_PRIVATE);
        return prefs.getString("code_verifier", null);
    }
    
    private void clearCodeVerifier() {
        android.content.SharedPreferences prefs = context.getSharedPreferences("XProvider", Context.MODE_PRIVATE);
        prefs.edit().remove("code_verifier").apply();
    }
}

File: ios/Sources/SocialLoginPlugin/XProvider.swift
import Foundation
import Capacitor
import AuthenticationServices
import CryptoKit
import Security

@objc(XProvider)
public class XProvider: NSObject {
    private static let TAG = "XProvider"
    private static let OAUTH_URL = "https://x.com/i/oauth2/authorize"
    
    private var clientId: String?
    private var redirectUrl: String?
    private var accessToken: String?
    private var refreshToken: String?
    private var webAuthSession: ASWebAuthenticationSession?
    
   func initialize(clientId: String, redirectUrl: String) {
        self.clientId = clientId
        self.redirectUrl = redirectUrl
    }
    
   func login(payload: [String: Any], completion: @escaping (Result<XLoginResponse, Error>) -> Void) {
        guard let clientId = clientId, let redirectUrl = redirectUrl else {
            completion(.failure(XProviderError.notInitialized))
            return
        }
        
        // Generate PKCE parameters
        let codeVerifier = generateCodeVerifier()
        let codeChallenge = generateCodeChallenge(codeVerifier: codeVerifier)
        let state = generateState()
        
        // Store code verifier for later use
        storeCodeVerifier(codeVerifier)
        
        // Build OAuth URL
        let scopes = payload["scopes"] as? String ?? "tweet.read users.read users.email"
        let authUrl = buildAuthUrl(codeChallenge: codeChallenge, state: state, scopes: scopes)
        
        // Start authentication session
        startAuthenticationSession(authUrl: authUrl, redirectUrl: redirectUrl) { [weak self] result in
            switch result {
            case .success(let code):
                self?.handleOAuthSuccess(code: code, completion: completion)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
   func logout() {
        // Clear code verifier if any
        clearCodeVerifier()
    }
    
   func getAuthorizationCode() -> String? {
        return nil // Not implemented for X provider - use login instead
    }
    
   func isLoggedIn() -> Bool {
        return false // For X provider, we don't store tokens locally - backend should handle validation
    }
    
    private func generateCodeVerifier() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let data = Data(bytes)
        return data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    private func generateCodeChallenge(codeVerifier: String) -> String {
        let data = codeVerifier.data(using: .utf8)!
        let hash = SHA256.hash(data: data)
        let hashData = Data(hash)
        return hashData.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    private func generateState() -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let data = Data(bytes)
        return data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
    
    private func buildAuthUrl(codeChallenge: String, state: String, scopes: String) -> String {
        var components = URLComponents(string: XProvider.OAUTH_URL)!
        components.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "redirect_uri", value: redirectUrl),
            URLQueryItem(name: "scope", value: scopes),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: codeChallenge),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        return components.url!.absoluteString
    }
    
    private func startAuthenticationSession(authUrl: String, redirectUrl: String, completion: @escaping (Result<String, Error>) -> Void) {
        guard let url = URL(string: authUrl) else {
            completion(.failure(XProviderError.invalidUrl))
            return
        }
        
        webAuthSession = ASWebAuthenticationSession(url: url, callbackURLScheme: URL(string: redirectUrl)?.scheme) { [weak self] callbackURL, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let callbackURL = callbackURL else {
                completion(.failure(XProviderError.noCallbackUrl))
                return
            }
            
            // Parse callback URL
            let components = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)
            let code = components?.queryItems?.first { $0.name == "code" }?.value
            let error = components?.queryItems?.first { $0.name == "error" }?.value
            
            if let error = error {
                completion(.failure(XProviderError.oauthError(error)))
                return
            }
            
            if let code = code {
                completion(.success(code))
            } else {
                completion(.failure(XProviderError.noAuthorizationCode))
            }
        }
        
        webAuthSession?.presentationContextProvider = self
        webAuthSession?.start()
    }
    
    private func handleOAuthSuccess(code: String, completion: @escaping (Result<XLoginResponse, Error>) -> Void) {
        guard let codeVerifier = getStoredCodeVerifier() else {
            completion(.failure(XProviderError.noCodeVerifier))
            return
        }
        
        // Clear code verifier
        clearCodeVerifier()
        
        // Return code and code_verifier for backend to exchange
        let response = XLoginResponse(
            accessToken: ["token": code, "code_verifier": codeVerifier],
            profile: [:]
        )
        
        completion(.success(response))
    }

    private func storeCodeVerifier(_ codeVerifier: String) {
        UserDefaults.standard.set(codeVerifier, forKey: "XProvider_code_verifier")
    }
    
    private func getStoredCodeVerifier() -> String? {
        return UserDefaults.standard.string(forKey: "XProvider_code_verifier")
    }
    
    private func clearCodeVerifier() {
        UserDefaults.standard.removeObject(forKey: "XProvider_code_verifier")
    }
}

extension XProvider: ASWebAuthenticationPresentationContextProviding {
    public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return UIApplication.shared.windows.first!
    }
}

// Response types
@objc(XLoginResponse)
public class XLoginResponse: NSObject {
    public let accessToken: [String: Any]
    public let profile: [String: Any]
    
    init(accessToken: [String: Any], profile: [String: Any]) {
        self.accessToken = accessToken
        self.profile = profile
    }
}

// Error types
enum XProviderError: Error, LocalizedError {
    case notInitialized
    case invalidUrl
    case noCallbackUrl
    case noAuthorizationCode
    case noCodeVerifier
    case noData
    case oauthError(String)
    
    var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "X Provider not initialized"
        case .invalidUrl:
            return "Invalid URL"
        case .noCallbackUrl:
            return "No callback URL received"
        case .noAuthorizationCode:
            return "No authorization code received"
        case .noCodeVerifier:
            return "No code verifier found"
        case .noData:
            return "No data received"
        case .oauthError(let error):
            return "OAuth error: \(error)"
        }
    }
} 